demo/demo/tests/test_paddle.py

```diff
@@ -17,6 +20,141 @@
+    # get the browser info token, which will be used later on.
+    browser_info = url.strip('/').split('/')[-1]
+    session = requests.Session()
+
+    # hit the redirect url, and start the payment procedure
+    payment_page = session.get(url, allow_redirects=True)
+
+    # the checkout id has been initialized.
+    payment_page_query = parse.parse_qs(parse.urlsplit(payment_page.url).query)
+    checkout_id = payment_page_query['checkout_id'][0]
+    assert 'checkout_id' in payment_page_query
+
+    # parse the react to get the needed urls
+    checkout_api_url = payment_page.text.split("REACT_APP_CHECKOUT_API_URL: '")[-1].split("',")[0]
+    ld_proxy_domain = payment_page.text.split("REACT_APP_LD_PROXY_URL: '")[-1].split("',")[0]
+    spreedly_api_url = payment_page.text.split("REACT_APP_SPREEDLY_API_URL: '")[-1].split("',")[0]
+
+    # Sanity checks on the variables
+    assert 'paddle' in checkout_api_url
+    checkout_api_url = f'{checkout_api_url}/checkout/{checkout_id}'
+    if not checkout_api_url.endswith('/'):
+        checkout_api_url += '/'
+
+    assert 'paddle' in ld_proxy_domain
+
+    assert 'spreedly' in spreedly_api_url
+    assert spreedly_api_url.endswith('v1'), 'Unexpected version of the spreedly API, this test might not work correctly'
+    if not spreedly_api_url.endswith('/'):
+        spreedly_api_url += '/'
+
+    # prepare the customer info request
+    customer_info_url = parse.urljoin(checkout_api_url, 'customer-info')
+    customer_info = requests.post(
+        customer_info_url,
+        json={
+            'data': {
+                'email': email,
+                'audience_optin': False,
+                'country_code': 'PL',
+                'postcode': 12345
+            }
+        }
+    )
+    payment_methods = customer_info.json().get(
+        'data', {}
+    ).get(
+        'available_payment_methods', []
+    )
+
+    # fetch the CARD payment method
+    for payment_method in payment_methods:
+        if payment_method['type'] == 'CARD':
+            assert 'spreedly_options' in payment_method
+            assert 'spreedly_environment_key' in payment_method['spreedly_options']
+            break
+    else:
+        assert False, 'Unexpected response from Paddle, payment method not found'
+
+    # Now we should have received the spreedly environment token.
+    spreedly_environment_key = payment_method['spreedly_options']['spreedly_environment_key']
+    assert spreedly_environment_key
+
+    # Set the payment method to CARD
+    set_payment_method_url = parse.urljoin(checkout_api_url, 'payment-method')
+    requests.patch(
+        set_payment_method_url,
+        json={
+            'data': {
+                "payment_method_type": "CARD"
+            }
+        }
+    )
+
+    # Send the card details to spreedly adn fetch the transaction token
+    spreedly_url = parse.urljoin(spreedly_api_url, f'payment_methods.json?environment_key={spreedly_environment_key}')
```

Comment:
Inconsistent. You're using `urljoin` to "join url parts the right way", however you're constructing query string with string interpolation instead of `urlencode`.

Reply:
:+1:

Reply:
Nice solution with `params`!